framework/src/vectorpostprocessors/ExtraIDIntegralVectorPostprocessor.C

```diff
@@ -57,53 +69,87 @@ ExtraIDIntegralVectorPostprocessor::ExtraIDIntegralVectorPostprocessor(
     p.resize(extra_ids.size());
     for (auto it : extra_ids)
       p[it.first] = it.second;
-    _var_extra_ids.push_back(&p);
+    _extra_ids.push_back(&p);
   }
 
-  // declare vectors containing integral values
+  // declare vectors containing variable integral values
   for (unsigned int i = 0; i < _nvar; ++i)
   {
     _vars.push_back(getVar("variable", i));
     _var_values.push_back(&coupledValue("variable", i));
     auto & p = declareVector(_vars[i]->name());
-    p.resize((*_var_extra_ids[0]).size());
-    _var_integrals.push_back(&p);
+    p.resize((*_extra_ids[0]).size());
+    _integrals.push_back(&p);
+  }
+
+  // declare vectors containing material property integral values
+  for (auto & name : _prop_names)
+  {
+    _props.push_back(&getMaterialPropertyByName<Real>(name));
+    auto & p = declareVector(name);
+    p.resize((*_extra_ids[0]).size());
+    _integrals.push_back(&p);
   }
 }
 
 void
 ExtraIDIntegralVectorPostprocessor::initialize()
 {
-  for (auto & var_integral : _var_integrals)
-    std::fill(var_integral->begin(), var_integral->end(), 0);
+  for (auto & integral : _integrals)
+    std::fill(integral->begin(), integral->end(), 0);
+
+  if (_average)
+    _volumes.assign((*_extra_ids[0]).size(), 0);
 }
 
 void
 ExtraIDIntegralVectorPostprocessor::execute()
 {
   if (hasBlocks(_current_elem->subdomain_id()))
   {
+    unsigned int i = 0;
     auto ipos = _unique_vpp_ids[_current_elem->id()];
-    for (unsigned int ivar = 0; ivar < _nvar; ++ivar)
+    for (unsigned int ivar = 0; ivar < _nvar; ++ivar, ++i)
       if (_vars[ivar]->hasBlocks(_current_elem->subdomain_id()))
         for (unsigned int qp = 0; qp < _qrule->n_points(); qp++)
-          (*_var_integrals[ivar])[ipos] += _JxW[qp] * _coord[qp] * (*_var_values[ivar])[qp];
+          (*_integrals[i])[ipos] += _JxW[qp] * _coord[qp] * (*_var_values[ivar])[qp];
+
+    for (unsigned int iprop = 0; iprop < _nprop; ++iprop, ++i)
+      if (hasBlockMaterialProperty<Real>(_prop_names[i]))
```

Comment:
This is a very expensive call that includes map lookups and searching through all active materials.

Can you look this up ahead of time in `initialize()`?

Reply:
Didn't know that. Thank you for pointing it out.

Reply:
That call eventually seems unnecessary.

Reply:
Have you tested it with block restriction?

Reply:
If block restricted, there could be an issue.
It would be better to add a loop for checking property existence in initialize() something like:
```c++
  for (const auto & id : blockIDs())
    for (const auto & prop_name : _prop_names)
    _fe_problem.hasBlockMaterialProperty(id, prop_name);
```

Reply:
It seemed that MOOSE is checking the blocks in `FEProblemBase::checkProblemIntegrity()`. If I don't properly assign blocks for materials, I get something like:

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'mat2', requested by 'integral' is not defined on block 1
Material property 'mat1', requested by 'integral' is not defined on block 2
```

Reply:
@NamjaeChoi  Thank you!  Good to know!!

Reply:
My initial thought was to use zero if a material property is not defined on a block, but it seems that such thing is not allowed.

Reply:
Perfect - looks good to me.